### PR TITLE
Use sonic-boom

### PR DESCRIPTION
**Pull request by mcollina**
This still needs some work but I've opened it get some feedbacks.

Fixes https://github.com/pinojs/pino/issues/341.
Fixes #393.

### CODE REVIEW

**Comment by mcollina** [comment]
This has a perf regression on extreme mode, I will investigate.

master:

```
benchBunyan*10000: 834.808ms
benchWinston*10000: 623.330ms
benchBole*10000: 336.978ms
benchDebug*10000: 368.957ms
benchLogLevel*10000: 464.211ms
benchPino*10000: 331.520ms
benchPinoExtreme*10000: 238.182ms
benchBunyan*10000: 715.397ms
benchWinston*10000: 538.094ms
benchBole*10000: 313.578ms
benchDebug*10000: 327.992ms
benchLogLevel*10000: 396.778ms
benchPino*10000: 306.623ms
benchPinoExtreme*10000: 217.418ms
```

This PR:

```
$ node benchmarks/basic.bench.js
benchBunyan*10000: 906.333ms
benchWinston*10000: 626.493ms
benchBole*10000: 330.269ms
benchDebug*10000: 352.773ms
benchLogLevel*10000: 422.380ms
benchPino*10000: 319.075ms
benchPinoExtreme*10000: 173.527ms
benchBunyan*10000: 734.764ms
benchWinston*10000: 536.962ms
benchBole*10000: 309.273ms
benchDebug*10000: 339.026ms
benchLogLevel*10000: 409.393ms
benchPino*10000: 318.122ms
benchPinoExtreme*10000: 150.465ms
```

File: package.json [hunk 1]
```diff
@@ -17,7 +17,7 @@
   ],
   "scripts": {
     "browser-test": "zuul tape test/browser*test.js --local",
-    "test": "standard | snazzy && NODE_OPTIONS='--no-warnings' tap --no-cov test/*test.js",
+    "test": "standard | snazzy && NODE_OPTIONS='--no-warnings' tap -j 4 --no-cov test/*test.js",
```
Comment (davidmarkclements) [inline package.json]: why not also parallelize ci (or do you only get one core on travis?)

File: pino.js [hunk 2]
```diff
@@ -382,6 +361,13 @@ function pino (opts, stream) {
 
 tools.defineLevelsProperty(pino)
 
+function extreme (dest) {
+  if (!dest) {
+    return new SonicBoom(process.stdout.fd)
```
Comment (davidmarkclements) [inline pino.js]: didn't we support a thing that allows for `process.stdout` and `process.stderr` to be the dest? 
Does `SonicBoom` support that, or only fd's?

File: test/fixtures/events/no-event-loop.js [hunk 3]
```diff
@@ -2,5 +2,5 @@ global.process = { __proto__: process, pid: 123456 }
 Date.now = function () { return 1459875739796 }
 require('os').hostname = function () { return 'abcdefghijklmnopqr' }
 var pino = require(require.resolve('./../../../'))
-var log = pino({extreme: true})
+var log = pino(pino.extreme())
```
Comment (davidmarkclements) [inline test/fixtures/events/no-event-loop.js]: why do we pass `pino.extreme` to `pino` here?.

**Comment by davidmarkclements** [comment]
doc updates will be separate PR? 

if so I don't think we should merge without updating docs (in case it gets forgotten)

File: pino.js [hunk 4]
```diff
@@ -286,15 +275,21 @@ function getPrettyStream (opts, prettifier) {
...
   var istream = stream
-  if (iopts && (iopts.writable || iopts._writableState)) {
+  if (iopts && (iopts.writable || iopts._writableState || iopts instanceof SonicBoom)) {
     istream = iopts
     iopts = defaultOptions
   }
   iopts = Object.assign({}, defaultOptions, iopts)
-  if (iopts.extreme && iopts.prettyPrint) throw Error('cannot enable pretty print in extreme mode')
+  if (iopts.extreme) {
+    throw new Error('the extreme option is deprecated, use require(\'pino\').extreme(dest) instead')
```
Comment (jsumners) [inline pino.js]: Replace "deprecated" with "removed".

File: pino.js [hunk 5]
```diff
@@ -286,15 +275,21 @@ function getPrettyStream (opts, prettifier) {
...
+  }
+  if (stream instanceof SonicBoom && iopts.prettyPrint) throw Error('cannot enable pretty print in extreme mode')
```
Comment (jsumners) [inline pino.js]: Is this true? Pretty printing now just consists of a `.write` to the stream.

File: test/fixtures/events/no-event-loop.js (hunk 3 above)
Comment (jsumners) [reply test/fixtures/events/no-event-loop.js]: That's the new API. `pino.extreme` is a utility method to create a destination stream.

**Comment by mcollina** [comment]
This PR:

```
$ node benchmarks/basic.bench.js
benchBunyan*10000: 849.859ms
benchWinston*10000: 631.939ms
benchBole*10000: 309.742ms
benchDebug*10000: 347.677ms
benchLogLevel*10000: 459.549ms
benchPino*10000: 304.755ms
benchPinoDestination*10000: 219.132ms
benchPinoExtreme*10000: 172.622ms
benchBunyan*10000: 727.658ms
benchWinston*10000: 501.120ms
benchBole*10000: 316.504ms
benchDebug*10000: 332.890ms
benchLogLevel*10000: 406.898ms
benchPino*10000: 295.664ms
benchPinoDestination*10000: 223.564ms
benchPinoExtreme*10000: 160.127ms
```

Master

```
$ node benchmarks/basic.bench.js
benchBunyan*10000: 843.972ms
benchWinston*10000: 610.107ms
benchBole*10000: 324.121ms
benchDebug*10000: 352.105ms
benchLogLevel*10000: 442.383ms
benchPino*10000: 331.777ms
benchPinoExtreme*10000: 167.160ms
benchBunyan*10000: 716.411ms
benchWinston*10000: 604.688ms
benchBole*10000: 376.154ms
benchDebug*10000: 381.350ms
benchLogLevel*10000: 426.684ms
benchPino*10000: 301.478ms
benchPinoExtreme*10000: 150.016ms
```

Note that there is the new `destination` object based on SonicBoom that gives us a nice boost.

File: docs/API.md [hunk 6]
```diff
@@ -609,6 +608,43 @@ Returns a unix time in seconds, like `,"time":1493426328`.
 Returns an empty string. This function is used when the `timestamp` option
 is set to `false`.
 
+<a id=".destination"></a>
+# .destination(dest?)
+
+Create a pino destination. It yields a 30% over using a standard Node.js
```
Comment (jsumners) [inline docs/API.md]: "It yields a stream with 30% more throughput than a standard Node.js stream."

or

"It returns a stream with significantly more throughput than a standard Node.js stream."

File: docs/API.md [hunk 7]
```diff
@@ -609,6 +608,43 @@ Returns a unix time in seconds, like `,"time":1493426328`.
...
+stream.
+
+```js
+const pino = require('pino')
+const logger = pino(pino.destination('./my-file'))
+const logger2 = pino(pino.destination())
+```
+
+`dest` could be both a file or a file descriptor. If it is omitted, it
```
Comment (jsumners) [inline docs/API.md]: "`dest` can be either a file or a file descriptor."

File: docs/API.md [hunk 8]
```diff
@@ -609,6 +608,43 @@ Returns a unix time in seconds, like `,"time":1493426328`.
...
+will be `process.stdout.fd`.
+
+The default `stream` is a destination.
+
+It is based on [`sonic-boom`](https://github.com/mcollina/sonic-boom).
+
+<a id=".extremie"></a>
```
Comment (jsumners) [inline docs/API.md]: A wayward `i` is here. Also, remove the leading `.`.

File: docs/API.md [hunk 9]
```diff
@@ -609,6 +608,43 @@ Returns a unix time in seconds, like `,"time":1493426328`.
...
+# .extreme(dest?)
+
+Create an extreme mode destination. It yields an additional 60% performance.
+There are trade-off's should be understood before usage.
```
Comment (jsumners) [inline docs/API.md]: "trade-offs that should be"

File: docs/API.md [hunk 10]
```diff
@@ -609,6 +608,43 @@ Returns a unix time in seconds, like `,"time":1493426328`.
...
+See [Extreme mode explained](extreme.md).
+
+```js
+const pino = require('pino')
+const logger = pino(pino.extreme('./my-file'))
+const logger2 = pino(pino.extreme())
+```
+
+`dest` could be both a file or a file descriptor. If it is omitted, it
```
Comment (jsumners) [inline docs/API.md]: "`dest` can be either a file or a file descriptor."

File: docs/API.md [hunk 11]
```diff
@@ -611,27 +611,28 @@ is set to `false`.
...
-`dest` could be both a file or a file descriptor. If it is omitted, it
+`dest` could be either a file or a file descriptor. If it is omitted, it
 will be `process.stdout.fd`.
 
 The default `stream` is a destination.
 
 It is based on [`sonic-boom`](https://github.com/mcollina/sonic-boom).
 
-<a id=".extremie"></a>
+<a id=".extreme"></a>
```
Comment (jsumners) [inline docs/API.md]: The `.` either needs to be removed here or added in the table of contents link.

**Comment by mcollina** [comment]
@jsumners PTAL

**Comment by mcollina** [comment]
@jsumners the events tests are not passing on Node 6 on Linux. However they are on Mac. Can you have a look?

```
 # Subtest: terminates on SIGHUP when no other handlers registered
        1..2
        not ok 1 - should be equal
          ---
          found: null
          wanted: 0
          compare: ===
          at:
            line: 80
            column: 7
            file: test/events.test.js
            type: ChildProcess
          stack: |
            ChildProcess.<anonymous> (test/events.test.js:80:7)
          source: |
            t.is(code, 0)
          ...
        
        not ok 2 - should not be equal
          ---
          found: null
          doNotWant: null
          compare: '!=='
          at:
            line: 84
            column: 7
            file: test/events.test.js
            type: ChildProcess
          stack: |
            ChildProcess.<anonymous> (test/events.test.js:84:7)
          source: |
            t.notEqual(output.match(/"msg":"h"/), null)
          ...
        
        # failed 2 of 2 tests
    not ok 4 - terminates on SIGHUP when no other handlers registered # time=1167.903ms
```

**Comment by jsumners** [comment]
I'll have to setup a Linux environment to test. That _could_ happen today, but we'll see.

**Comment by mcollina** [comment]
For some reasons I thought you had one. Anyway, thanks ;). I'm pretty snowed under, and I can't take this over the finishing line.

**Comment by jsumners** [comment]
Not a development environment. But it doesn't take much to get one going. I just a bit busy right now.

File: docs/API.md [hunk 12]
```diff
@@ -609,6 +608,44 @@ Returns a unix time in seconds, like `,"time":1493426328`.
...
+const logger = pino(pino.destination('./my-file'))
+const logger2 = pino(pino.destination())
+```
+
+`dest` could be either a file or a file descriptor. If it is omitted, it
+will be `process.stdout.fd`.
+
+The default `stream` is a destination.
+
+It is based on [`sonic-boom`](https://github.com/mcollina/sonic-boom).
```
Comment (davidmarkclements) [inline docs/API.md]: Pino destinations are based on

or 

The `pino.destination` API is implemented on `sonic-boom`

File: docs/API.md [hunk 13]
```diff
@@ -609,6 +608,44 @@ Returns a unix time in seconds, like `,"time":1493426328`.
...
+
+<a id="extreme"></a>
+# .extreme(dest?)
+
+Create an extreme mode destination. It yields an additional 60% performance.
```
Comment (davidmarkclements) [inline docs/API.md]: This yields

Comment (davidmarkclements) [reply docs/API.md]: 60% performance boost

File: docs/API.md [hunk 14]
```diff
@@ -609,6 +608,44 @@ Returns a unix time in seconds, like `,"time":1493426328`.
...
+```js
+const pino = require('pino')
+const logger = pino(pino.extreme('./my-file'))
+const logger2 = pino(pino.extreme())
+```
+
+`dest` can be either a file or a file descriptor. If it is omitted, it
+will be `process.stdout.fd`.
+
+It is based on [`sonic-boom`](https://github.com/mcollina/sonic-boom).
```
Comment (davidmarkclements) [inline docs/API.md]: Pino destinations are based on

or

The pino.destination API is implemented on sonic-boom

**Pull request merged**